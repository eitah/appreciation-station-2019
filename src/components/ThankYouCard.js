import React, { Component } from "react";

const CANVAS_CONFIG = {
  breakPoint: 720,
  smallFontSize: 24,
  largeFontSize: 48,
  lineSpacing: 10,
  smallSpacing: 20,
  largeSpacing: 60
};

export default class ThankYouCard extends Component {
  constructor(props) {
    super(props);
    this.canvasRef = React.createRef();
    this.imageRef = React.createRef();
  }

  componentDidMount() {
    this.createImageForCanvas();
    this.setImageSrc();
    this.imageRef.current.addEventListener("load", this.drawCanvas, false);
    window.addEventListener("resize", this.drawCanvas);
  }

  componentDidUpdate() {
    this.setImageSrc();
  }

  componentWillUnmount() {
    this.imageRef.current.removeEventListener("load", this.drawCanvas);
    window.removeEventListener("resize", this.drawCanvas);
  }

  createImageForCanvas = () => {
    var img = new Image();
    this.imageRef.current = img;
    img.crossOrigin = "anonymous";
    return img;
  };

  setCanvasSize = (canvas, width, height) => {
    canvas.width = width;
    canvas.height = height;
  };

  setImageSrc = () => {
    const img = this.imageRef.current;
    img.src = this.props.backgroundImage.file.url;
  };

  drawBackgroundImage = (context, width, height) => {
    const img = this.imageRef.current;
    context.drawImage(img, 0, 0, width, height);
  };

  drawText = (context, width, height) => {
    const canvasMessage = this.getCanvasMessage();
    const textStyle = this.getTextStyleConfig(width);

    context.font = textStyle.font;
    context.fillStyle = textStyle.fillStyle;

    const lines = this.getLines(context, canvasMessage, textStyle.maxWidth);
    lines.forEach((line, index) => {
      const position = this.calculateTextPosition(
        index,
        lines.length,
        textStyle.lineHeight,
        textStyle.horizontalSpacing,
        textStyle.verticalSpacing,
        height
      );
      context.fillText(line, position.left, position.top, textStyle.maxWidth);
    });
  };

  getCanvasMessage = () => {
    let canvasMessage = this.props.message.text.replace(
      "{NAME}",
      this.props.member.name
    );

    const singleQuote = String.fromCharCode(8217);

    canvasMessage = canvasMessage.replace("'", singleQuote);

    return canvasMessage;
  };

  getTextStyleConfig = width => {
    const fontSize =
      width < CANVAS_CONFIG.breakPoint
        ? CANVAS_CONFIG.smallFontSize
        : CANVAS_CONFIG.largeFontSize;

    const lineHeight = fontSize + CANVAS_CONFIG.lineSpacing;

    const horizontalSpacing =
      width < CANVAS_CONFIG.breakPoint
        ? CANVAS_CONFIG.smallSpacing
        : CANVAS_CONFIG.largeSpacing;

    const verticalSpacing =
      width < CANVAS_CONFIG.breakPoint
        ? CANVAS_CONFIG.smallSpacing
        : CANVAS_CONFIG.largeSpacing;

    const maxWidth = width * this.props.backgroundImage.maxTextWidth;

    const font = `600 ${fontSize}px Merriweather`;
    const fillStyle = "#ffffff";

    return {
      fontSize,
      lineHeight,
      horizontalSpacing,
      verticalSpacing,
      maxWidth,
      font,
      fillStyle
    };
  };

  getLines = (context, text, maxWidth) => {
    var words = text.split(" ");
    var lines = [];
    var currentLine = words[0];

    for (var i = 1; i < words.length; i++) {
      var word = words[i];
      var width = context.measureText(currentLine + " " + word).width;
      if (width < maxWidth) {
        currentLine += " " + word;
      } else {
        lines.push(currentLine);
        currentLine = word;
      }
    }
    lines.push(currentLine);
    return lines;
  };

  calculateTextPosition = (
    currentIndex,
    numOfLines,
    lineHeight,
    horizontalSpacing,
    verticalSpacing,
    height
  ) => {
    if (this.props.backgroundImage.textPosition === "topLeft") {
      const multiplier = currentIndex + 1;
      const space = multiplier * lineHeight + verticalSpacing / 2;
      return {
        left: horizontalSpacing,
        top: space
      };
    } else if (this.props.backgroundImage.textPosition === "bottomLeft") {
      const multiplier = numOfLines - 1 - currentIndex;
      const space = multiplier * lineHeight + verticalSpacing;

      return {
        left: horizontalSpacing,
        top: height - space
      };
    }
  };

  drawCanvas = () => {
    const canvas = this.canvasRef.current;
    const container = canvas.parentNode;
    const width = container.offsetWidth;
    const height = width * (9 / 16); // maintain 16:9 aspect ratio

    this.setCanvasSize(canvas, width, height);

    const context = canvas.getContext("2d");

    this.drawBackgroundImage(context, width, height);
    this.drawText(context, width, height);
  };

  download = e => {
    const canvas = this.canvasRef.current;
    const imgData = canvas.toDataURL("png");
    e.target.href = imgData;
  };

  render() {
    return (
      <React.Fragment>
        <div className="ThankYouCardContainer">
          {this.props.prevArrow()}
          <div className="ThankYouCard">
            <canvas ref={this.canvasRef} />
          </div>
          {this.props.nextArrow()}
        </div>
        <div>
          <a
            className="Share Button"
            href="#"
            onClick={this.download}
            download={`Thank you ${
              this.props.member.name
            }! - ${new Date().toISOString()}.png`}
          >
            Download
          </a>
        </div>
      </React.Fragment>
    );
  }
}
