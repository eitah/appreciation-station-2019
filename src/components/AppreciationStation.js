import React, { useState, useEffect } from "react";
import { AppreciationStationConfig } from "../lib/data";
import ThankYouCard from "./ThankYouCard";
import WeDidIt from "../assets/images/WeDidIt.svg";
import PreviousNextArrow from "../assets/images/PreviousNextArrow.svg";

export default function AppreciationStation() {
  const [thankYous, setThankYous] = useState([]);
  const [searchName, setSearchName] = useState("");
  const [nameNotFound, setNameNotFound] = useState(false);
  const [
    currentVisibleThankYouIndex,
    setCurrentVisibleThankYouIndex
  ] = useState(0);

  useEffect(() => {
    function createThankYous(
      receivedImages,
      receivedMessages,
      receivedMembers
    ) {
      const mentorMessages = receivedMessages.filter(msg => msg.isMentor);
      const menteeMessages = receivedMessages.filter(msg => msg.isMentee);
      const speakerMessages = receivedMessages.filter(msg => msg.isSpeaker);
      const sponsorMessages = receivedMessages.filter(msg => msg.isSponsor);

      return receivedMembers.map(member => {
        let messages;
        if (member.isMentor) messages = mentorMessages;
        if (member.isMentee) messages = menteeMessages;
        if (member.isSpeaker) messages = speakerMessages;
        if (member.isSponsor) messages = sponsorMessages;

        return {
          member: member,
          backgroundImage: getRandom(receivedImages),
          message: getRandom(messages)
        };
      });
    }

    AppreciationStationConfig.getBackgroundImages().then(imagesResponse => {
      AppreciationStationConfig.getMemberMessages().then(messagesResponse => {
        AppreciationStationConfig.getMembers().then(membersResponse => {
          const localThankYous = createThankYous(
            imagesResponse,
            messagesResponse,
            membersResponse
          );
          shuffle(localThankYous, setThankYous);
        });
      });
    });
  }, []);

  // Fisher-Yates shuffle
  function shuffle(collection, collectionSetter) {
    let array = [...collection];
    for (let i = array.length - 1; i > 0; i--) {
      let j = Math.floor(Math.random() * (i + 1)); // random index from 0 to i
      [array[i], array[j]] = [array[j], array[i]]; // swap elements
    }
    collectionSetter(array);
  }

  function getRandom(array) {
    return array[Math.floor(Math.random() * array.length)];
  }

  function search(e) {
    e.preventDefault();

    const indexToBeVisible = thankYous.findIndex(ty =>
      ty.member.name.toLowerCase().includes(searchName.toLowerCase())
    );
    if (indexToBeVisible >= 0) {
      setCurrentVisibleThankYouIndex(indexToBeVisible);
    } else {
      setNameNotFound(true);
    }
  }

  function renderPrevArrow() {
    const isHiddenClassName =
      currentVisibleThankYouIndex <= 0 ? "is-hidden" : "";
    const className = `Prev is-clickable ${isHiddenClassName}`;

    return (
      <img
        src={PreviousNextArrow}
        className={className}
        onClick={() =>
          setCurrentVisibleThankYouIndex(currentVisibleThankYouIndex - 1)
        }
        alt="Go to previous thank you card"
      />
    );
  }

  function renderNextArrow() {
    const isHiddenClassName =
      currentVisibleThankYouIndex >= thankYous.length - 1 ? "is-hidden" : "";
    const className = `Next is-clickable ${isHiddenClassName}`;
    return (
      <img
        src={PreviousNextArrow}
        className={className}
        onClick={() =>
          setCurrentVisibleThankYouIndex(currentVisibleThankYouIndex + 1)
        }
        alt="Go to next thank you card"
      />
    );
  }

  const shouldShowThankYouCard = thankYous && thankYous.length > 0;

  return (
    <section className="Section">
      <img className="BackgroundHeading" src={WeDidIt} alt="We Did It text" />
      <h2 className="h1 Section-heading">Appreciation Station</h2>
      <div className="Grid Grid--noBottomSpacing">
        <div className="GridItem GridItem--full">
          <p className="Subtitle">
            Did you have a great pairing experience and want to let your pair
            know you appreciate them? Did you have a great takeaway from one of
            our lightning talks? Search for their name and send them a thank you
            note. We've had 42 different mentors, 66 different mentees, 16
            different speakers, and 11 different sponsors this year. That's a
            lot of people to thank!
          </p>
        </div>
      </div>
      <div>
        <form onSubmit={search}>
          <label htmlFor="thankYouSearch">
            <input
              type="text"
              className="SearchInput"
              placeholder="Search for someone you'd like to thank"
              onChange={e => setSearchName(e.target.value)}
            />
          </label>
          <button type="submit" className="SearchButton">
            Search
          </button>
        </form>
      </div>
      {nameNotFound && (
        <div className="Grid Grid--topSpacing Grid--noBottomSpacing">
          <div className="GridItem GridItem--full">
            <p>
              <strong>
                Whoops, looks like we don't know anyone with that name. Check
                your spelling and try again.
              </strong>
            </p>
          </div>
        </div>
      )}
      {shouldShowThankYouCard && (
        <ThankYouCard
          member={thankYous[currentVisibleThankYouIndex].member}
          message={thankYous[currentVisibleThankYouIndex].message}
          backgroundImage={
            thankYous[currentVisibleThankYouIndex].backgroundImage
          }
          prevArrow={renderPrevArrow}
          nextArrow={renderNextArrow}
        />
      )}
    </section>
  );
}
