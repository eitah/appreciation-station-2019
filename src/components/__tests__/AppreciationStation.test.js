import React from "react";
import { shallow } from "enzyme";
import AppreciationStation from "../AppreciationStation";
import ThankYouCard from "../ThankYouCard";

const mockThankYous = [
  {
    member: {
      name: "Test member 1",
      isMentor: true
    },
    message: {
      text: "Thank you {NAME} for being a great mentor!"
    },
    backgroundImage: {
      file: {
        url: "http://testurl1.com"
      }
    }
  },
  {
    member: {
      name: "Test member 2",
      isMentee: true
    },
    message: {
      text: "Thank you {NAME} for being a great mentee!"
    },
    backgroundImage: {
      file: {
        url: "http://testurl2.com"
      }
    }
  },
  {
    member: {
      name: "Test member 3",
      isSpeaker: true
    },
    message: {
      text: "Thank you {NAME} for being a great speaker!"
    },
    backgroundImage: {
      file: {
        url: "http://testurl3.com"
      }
    }
  },
  {
    member: {
      name: "Test member 4",
      isSponsor: true
    },
    message: {
      text: "Thank you {NAME} for being a great sponsor!"
    },
    backgroundImage: {
      file: {
        url: "http://testurl4.com"
      }
    }
  }
];

let subject;

describe("AppreciationStation", () => {
  beforeEach(() => {
    subject = shallow(<AppreciationStation thankYous={mockThankYous} />);
  });

  describe("#render", () => {
    it("matches snapshot", () => {
      expect(subject).toMatchSnapshot();
    });
  });

  describe("#search", () => {
    describe("when name is not found", () => {
      it("renders a ThankYouCard component", () => {
        subject.find("form").simulate("submit", { preventDefault: jest.fn() });
        expect(subject).toMatchSnapshot();
        expect(subject.find("strong")).toHaveLength(1);
      });
    });
  });
});
