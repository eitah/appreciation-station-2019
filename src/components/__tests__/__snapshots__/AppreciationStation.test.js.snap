// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppreciationStation #render matches snapshot 1`] = `
<Fragment>
  <div>
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="thankYouSearch"
      >
        <span
          className="SearchInput-label"
        >
          Search for someone you'd like to thank
        </span>
        <input
          className="SearchInput"
          id="thankYouSearch"
          onChange={[Function]}
          placeholder="Search for name"
          type="text"
        />
      </label>
      <button
        className="SearchButton"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <ThankYouCard
    backgroundImage={
      Object {
        "file": Object {
          "url": "http://testurl1.com",
        },
      }
    }
    member={
      Object {
        "isMentor": true,
        "name": "Test member 1",
      }
    }
    message={
      Object {
        "text": "Thank you {NAME} for being a great mentor!",
      }
    }
    nextArrow={[Function]}
    prevArrow={[Function]}
  />
</Fragment>
`;

exports[`AppreciationStation #search when name is not found renders a ThankYouCard component 1`] = `
<Fragment>
  <div>
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="thankYouSearch"
      >
        <span
          className="SearchInput-label"
        >
          Search for someone you'd like to thank
        </span>
        <input
          className="SearchInput"
          id="thankYouSearch"
          onChange={[Function]}
          placeholder="Search for name"
          type="text"
        />
      </label>
      <button
        className="SearchButton"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <ThankYouCard
    backgroundImage={
      Object {
        "file": Object {
          "url": "http://testurl1.com",
        },
      }
    }
    member={
      Object {
        "isMentor": true,
        "name": "Test member 1",
      }
    }
    message={
      Object {
        "text": "Thank you {NAME} for being a great mentor!",
      }
    }
    nextArrow={[Function]}
    prevArrow={[Function]}
  />
</Fragment>
`;
